primary/src/grpc_server.rs

```diff
@@ -0,0 +1,182 @@
+
+            self.tx_get_block_commands
+                .send(BlockCommand::GetBlock {
+                    id: collection_id,
+                    sender: tx_get_block,
+                })
+                .await
+                .unwrap();
+
+            let timer = sleep(Duration::from_millis(5_000));
+            tokio::pin!(timer);
+
+            tokio::select! {
+                Ok(result) = rx_get_block => {
+                    get_blocks_response = match result {
+                        Ok(block) => {
+                            let mut retrieval_results = vec![];
+                            for batch in block.batches {
+                                retrieval_results.push(CollectionRetrievalResult{
+                                    retrieval_result: Some(types::RetrievalResult::Batch(BatchMessageProto::from(batch)))
+                                });
+                            }
+
+                            Ok(GetCollectionsResponse {
+                                result: retrieval_results,
+                            })
+                        },
+                        Err(err) => {
+                            Ok(GetCollectionsResponse {
+                                result: vec![CollectionRetrievalResult{
+                                    retrieval_result: Some(types::RetrievalResult::Error(err.into())),
+                                }]
+                            })
+                        }
+                    };
+                },
+                () = &mut timer => {
+                    get_blocks_response = Err(Status::internal("Timeout, no result has been received in time"));
+                }
+            }
+        } else if !collection_ids.is_empty() {
+            // Get multiple blocks
+            let tx_get_blocks: oneshot::Sender<Result<GetBlocksResponse, BlocksError>>;
+            let rx_get_blocks: oneshot::Receiver<Result<GetBlocksResponse, BlocksError>>;
+            (tx_get_blocks, rx_get_blocks) = oneshot::channel();
+
+            let ids = collection_ids
+                .iter()
+                .map(|c_id| CertificateDigest::new(c_id.f_bytes.clone().try_into().unwrap()))
+                .collect();
+
+            self.tx_get_block_commands
+                .send(BlockCommand::GetBlocks {
+                    ids,
+                    sender: tx_get_blocks,
+                })
+                .await
+                .unwrap();
+
+            let timer = sleep(Duration::from_millis(5_000));
+            tokio::pin!(timer);
+
+            tokio::select! {
+                Ok(result) = rx_get_blocks => {
+                    get_blocks_response = match result {
+                        Ok(blocks_response) => {
+                            let mut retrieval_results = vec![];
+                            for block_result in blocks_response.blocks {
+                                match block_result {
```

Comment:
It sounds like we want an auxiliary `Fn Result<GetBlockSReponse> -> CollectionRetrievalResult` then we can `retrieval_results.push (foo(block_result))`.